helix-term/src/ui/editor.rs

```diff
@@ -152,6 +154,45 @@ impl EditorView {
         self.render_statusline(doc, view, statusline_area, surface, theme, is_focused);
     }
 
+    pub fn doc_set_indent_guides(doc: &mut Document, offset: Position, height: u16, theme: &Theme) {
+        const INDENT_GUIDE_GROUP: &str = "indent-guides";
+        doc.clear_text_annotations(INDENT_GUIDE_GROUP);
+
+        let tab_width = doc.tab_width();
+        // Avoid lifetime issues by cloning the rope, which is cheap
+        let text = doc.text().clone();
+        let annots = text
+            .lines_at(offset.row)
+            .take(height as usize)
+            .scan(0, |prev_level, line| {
+                let level = indent_level_for_line(line, tab_width);
+                let line_is_blank = line.chunks().all(|s| s.trim().is_empty());
+
+                // Use the previous line's indent guide if current line is blank.
+                // This prevents breaks in the guides on newlines.
+                if line_is_blank && *prev_level != 0 {
+                    Some(*prev_level)
+                } else {
+                    *prev_level = level;
+                    Some(level)
+                }
+            })
+            .zip(offset.row..)
+            .flat_map(|(level, line_no)| {
+                let positions = (0..level).map(|step| step * tab_width);
+                positions.map(move |pos| TextAnnotation {
+                    text: "│".into(),
+                    style: theme
+                        .try_get("ui.indent-guide")
+                        .unwrap_or_else(|| theme.get("comment")),
+                    line: line_no,
+                    kind: TextAnnotationKind::Overlay(pos),
+                })
+            });
+
+        doc.push_text_annotations(INDENT_GUIDE_GROUP, annots);
+    }
```

Comment:
It seems a bit inefficient for me to do this via indent guides, where you have to pre-allocate and calculate these markers for each line.

I would directly modify the rendering code in a similar manner to https://github.com/helix-editor/helix/pull/1802

where you track the indent level as you render and appropriately substitute

Reply:
Done in 970d6a8 \o/